### ISSUE TITLE
Sum issues with FP32

### ISSUE THREAD

**Issue by emcastillo** [pytorch/pytorch#59863]
## 🐛 Bug

When doing the sum of a large array of fp32 in the cpu we are prone to have errors due to rounding when using multithreading.

## To Reproduce

1. export OMP_NUM_THREADS=2
1. torch.tensor((np.arange(8 * 80 * 80 * 80) - 4 * 80 * 80 * 80 + 0.5)).to( torch.float32)

## Expected behavior

1. Result is `131072.` although it should be 0.

## Environment

 - PyTorch Version (e.g., 1.0): master branch
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): python setup.py develop
 - Python version: 3.8
 - CUDA/cuDNN version: NA
 - GPU models and configuration: NA
 - Any other relevant information: NA

## Additional context

There is a two-step reduction going on. The first pass computes the following 2 values when running with two threads;
-2.09715e+12 and 0 2.09715e+12. However, the addition of these two results in 131072 due to round-off error.

In this case, the problem is that we are chunking the array according to the number of threads and do a regular sum at the end. So even though the sum kernel in the cpu was changed to reduce error, this only has an intra-thread effect.

Since the last step of this reduction is very small compared to the first one, would it be reasonable to do this only one by promoting the 
type to double when possible?

Thanks

**Comment by ezyang** [comment]
@peterbell10 are you at all interested in looking at this?

**Comment by peterbell10** [comment]
>In this case, the problem is that we are chunking the array according to the number of threads and do a regular sum at the end. So even though the sum kernel in the cpu was changed to reduce error, this only has an intra-thread effect.

This isn't an accurate characterization. Adding partial sums from chunked sections of the array is the exact same principle that makes cascade summation accurate. So in fact the two pass sum is equivalent to adding an extra level of accumulator into the cascade sum algorithm, which in general makes it _more accurate_ than single threaded. 

### So what's going wrong here? 

>There is a two-step reduction going on. The first pass computes the following 2 values when running with two threads;
-2.09715e+12 and 2.09715e+12.

In hex-float notation, the partial sums are:
```
-0x1.e847fep+40
       +
 0x1.e84800p+40
```
So, the error here is the difference between `e` and `10`, or error only in the 2 least significant bits.

Also note that the data being summed here is kind of a worst-case scenario: one chunk is sorted largest magnitude to smallest magnitude (the most rounding loss) and the second is sorted smallest to largest magnitude (the least rounding loss). If I instead shuffle the data first, the same sum returns -160. Funnily enough, the single threaded case gets exactly zero by way of more errors. It's just that the errors are perfectly symmetric and cancel out in the first and second half of the data.

### Conclusion
There is nothing wrong with multithreading here, and the cascade sum algorithm is working as intended. There are simply limits to the accuracy of a 32-bit float. 

I can take a look into quantifying the performance hit from accumulating in double precision later this week. And I do have some ideas for hybrid float-double summation that might reduce that hit.

**Comment by emcastillo** [comment]
Thanks for the detailed explanation! :)
